File: Week_07/212/word_search_ii.go
package main

type Trie struct {
	children map[byte]*Trie
	isEnd    bool
}

func findWords(board [][]byte, words []string) []string {
	res := make([]string, 0)
	trie := &Trie{children: map[byte]*Trie{}}
	m, n := len(board), len(board[0])
	for _, word := range words {
		node := trie
		for i := 0; i < len(word); i++ {
			if _, ok := node.children[word[i]]; !ok {
				node.children[word[i]] = &Trie{children: map[byte]*Trie{}}
			}
			node = node.children[word[i]]
		}
		node.isEnd = true
	}
	offset := [4][2]int{{1, 0}, {0, 1}, {-1, 0}, {0, -1}}
	var dfs func(i, j int, node *Trie, s string)
	dfs = func(i, j int, node *Trie, s string) {
		v := board[i][j]
		s += string(v)
		if _, ok := node.children[v]; !ok {
			return
		}
		node = node.children[v]
		if node.isEnd {
			res = append(res, s)
			node.isEnd = false
		}
		board[i][j] = '#'
		for k := 0; k < 4; k++ {
			dx, dy := offset[k][0], offset[k][1]
			x, y := i+dx, j+dy
			if x >= 0 && x < m && y >= 0 && y < n {
				dfs(x, y, node, s)
			}
		}
		board[i][j] = v

	}

	for i := 0; i < m; i++ {
		for j := 0; j < n; j++ {
			dfs(i, j, trie, "")
		}
	}
	return res
}

File: Week_07/212/word_search_ii.py
from typing import List


class Solution:
    def findWords(self, board: List[List[str]], words: List[str]) -> List[str]:
        trie = {}
        end_mark = '#'
        tmp_mark = '@'
        for word in words:
            t = trie
            for w in word:
                if w not in t:
                    t[w] = {}
                t = t[w]
            t[end_mark] = {}

        res = []
        m, n = len(board), len(board[0])

        def dfs(x: int, y: int, node: dict, s: str) -> None:
            v = board[x][y]
            if v not in node:
                return

            s += v
            node = node[v]
            if end_mark in node:
                res.append(s)
                node.pop(end_mark)

            board[x][y] = tmp_mark
            for dx, dy in [(1, 0), (-1, 0), (0, 1), (0, -1)]:
                x_, y_ = x + dx, y + dy
                if 0 <= x_ < m and 0 <= y_ < n:
                    dfs(x_, y_, node, s)
            board[x][y] = v

        for i in range(m):
            for j in range(n):
                dfs(i, j, trie, "")
        return res

    def test(self):
        board = [
            ['o', 'a', 'a', 'n'],
            ['e', 't', 'a', 'e'],
            ['i', 'h', 'k', 'r'],
            ['i', 'f', 'l', 'v']
        ]

        words = ["oath", "pea", "eat", "rain", "eate", "eie", "a"]

        ans = self.findWords(board, words)
        print(ans)


if __name__ == "__main__":
    Solution().test()
